File: xaml_transform3dparallax/CPP/ParallaxBackgroundHub.cpp
//
// ParallaxBackgroundHub.cpp
// Implementation of the ParallaxBackgroundHub class.
//

#include "pch.h"
#include "ParallaxBackgroundHub.h"

using namespace Transform3DParallax;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Documents;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Media::Media3D;

// The Templated Control item template is documented at http://go.microsoft.com/fwlink/?LinkId=234235

ParallaxBackgroundHub::ParallaxBackgroundHub()
{
	DefaultStyleKey = "Transform3DParallax.ParallaxBackgroundHub";
}


void ParallaxBackgroundHub::OnApplyTemplate()
{
	Hub::OnApplyTemplate();

	m_scrollViewer = dynamic_cast<ScrollViewer^>(GetTemplateChild(L"ScrollViewer"));
	m_backgroundTransform = dynamic_cast<CompositeTransform3D^>(GetTemplateChild(L"BackgroundTransform"));

	if (m_scrollViewer != nullptr && m_backgroundTransform != nullptr)
	{
		m_scrollViewer->SizeChanged += ref new Windows::UI::Xaml::SizeChangedEventHandler(this, &Transform3DParallax::ParallaxBackgroundHub::OnSizeChanged);
	}
}

void ParallaxBackgroundHub::UpdateForSizeChanged()
{
	// Since the background Border will always start in the top left, we don't need to use
	// the TransformToVisual() trick here.

	auto horizontalOffset = m_scrollViewer->ViewportWidth / 2.0;
	auto verticalOffset = m_scrollViewer->ViewportHeight / 2.0;

	m_backgroundTransform->CenterX = horizontalOffset;
	m_backgroundTransform->CenterY = verticalOffset;
}


void ParallaxBackgroundHub::OnSizeChanged(Object ^sender, SizeChangedEventArgs ^e)
{
	UpdateForSizeChanged();
}

File: Samples/XamlTransform3DParallax/cs/ParallaxBackgroundHub.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Documents;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Media3D;

// The Templated Control item template is documented at http://go.microsoft.com/fwlink/?LinkId=234235

namespace Transform3DParallax
{
    public sealed class ParallaxBackgroundHub : Hub
    {
        private ScrollViewer _scrollViewer;
        private CompositeTransform3D _backgroundTransform;

        public ParallaxBackgroundHub()
        {
            this.DefaultStyleKey = typeof(ParallaxBackgroundHub);
        }

        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            _scrollViewer = GetTemplateChild("ScrollViewer") as ScrollViewer;
            _backgroundTransform = GetTemplateChild("BackgroundTransform") as CompositeTransform3D;

            if (_scrollViewer != null && _backgroundTransform != null)
            {
                _scrollViewer.SizeChanged += ScrollViewer_SizeChanged;
            }
        }

        private void UpdateForSizeChanged()
        {
            // Since the background Border will always start in the top left, we don't need to use
            // the TransformToVisual() trick here.

            var horizontalOffset = _scrollViewer.ViewportWidth / 2.0;
            var verticalOffset = _scrollViewer.ViewportHeight / 2.0;

            _backgroundTransform.CenterX = horizontalOffset;
            _backgroundTransform.CenterY = verticalOffset;
        }

        private void ScrollViewer_SizeChanged(object sender, SizeChangedEventArgs e)
        {
            UpdateForSizeChanged();
        }
    }
}
